File: NewsExpress/app/src/main/java/com/example/newsexpress/MainActivity.java
package com.example.newsexpress;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import com.android.volley.AuthFailureError;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;
import com.example.newsexpress.adapter.NewsRecyclerViewAdapter;
import com.example.newsexpress.model.NewsModel;
import com.example.newsexpress.retrofit.Api;
import com.example.newsexpress.retrofit.ApiService;
import com.example.newsexpress.retrofit.RetrofitModel;
import com.example.newsexpress.viewmodel.NewsViewModel;

import org.json.JSONArray;
import org.json.JSONException;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private NewsRecyclerViewAdapter newsRecyclerViewAdapter;
    private NewsViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        init();

        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        newsRecyclerViewAdapter = new NewsRecyclerViewAdapter();
        recyclerView.setAdapter(newsRecyclerViewAdapter);

        viewModel = new ViewModelProvider(this).get(NewsViewModel.class);
        viewModel.getNewsItems().observe(this, newsItems ->{
            newsRecyclerViewAdapter.setNewsItems(newsItems);
        } );

        viewModel.fetchNewsData();
    }

    public void init(){
        recyclerView = findViewById(R.id.recyclerView);
    }



}

File: NewsExpresss/app/src/main/java/com/example/newsexpresss/MainActivity.kt
package com.example.newsexpresss

import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.newsexpresss.adpater.RecyclerViewAdapter
import com.example.newsexpresss.model.NewsModel
import com.example.newsexpresss.viewmodel.MainViewModel

class MainActivity : AppCompatActivity() {

//    private lateinit var recyclerView: RecyclerView
//    private lateinit var recyclerViewAdapter: RecyclerViewAdapter

    private lateinit var mainViewModel: MainViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)


        val newsArray = ArrayList<NewsModel>()

        val recyclerview = findViewById<RecyclerView>(R.id.recyclerView)
        recyclerview.layoutManager = LinearLayoutManager(this)

        mainViewModel = ViewModelProvider(this).get(MainViewModel::class.java)
        mainViewModel.newsItem.observe( this, Observer{
            newsArray.clear()
            for(news in it){
                newsArray.add(NewsModel(news.urlToImage.toString(),news.title.toString()))
                Log.d("size",newsArray.size.toString())
            }
        })

        mainViewModel.fetchNewsItems("us")

        val adapter = RecyclerViewAdapter(newsArray)
        recyclerview.adapter = adapter

    }

//
//    private fun init(){
//        recyclerView = findViewById(R.id.recyclerView)
//    }


}